front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,918 @@
+          this.handleNgeNodeOperation(ngeEvent.type, ngeEvent.node, netzgrafikDto);
+        },
+      });
+    };
+  }
+
+  public async handleNgeNodeOperation(
+    type: NGEEvent['type'],
+    node: NodeDto,
+    netzgrafikDto: NetzgrafikDto
+  ) {
+    const indexNode = this.getNodeById(node.id);
+    switch (type) {
+      case 'create':
+      case 'update': {
+        if (indexNode) {
+          if (indexNode.saved) {
+            // Update the key if trigram has changed and key is based on it
+            let nodeKey = indexNode.node.path_item_key;
+            if (
+              nodeKey.startsWith('trigram:') &&
+              indexNode.node.trigram !== node.betriebspunktName
+            ) {
+              nodeKey = `trigram:${node.betriebspunktName}`;
+            }
+            await this.apiUpdateNode({
+              ...indexNode.node,
+              ...OsrdNgeSync.castNgeNode(node, netzgrafikDto.labels),
+              id: indexNode.node.id,
+              path_item_key: nodeKey,
+            });
+          } else {
+            const newNode = {
+              ...indexNode.node,
+              ...OsrdNgeSync.castNgeNode(node, netzgrafikDto.labels),
+            };
+            // Create the node
+            await this.apiCreateNode(newNode);
+            // keep track of the ID given by NGE
+            this.nodesIdToKey[node.id] = newNode.path_item_key;
+          }
+        } else {
+          // It's an unknown node, we need to create it in the db
+          // We assume that `betriebspunktName` is a trigram
+          const key = `trigram:${node.betriebspunktName}`;
+          // Create the node
+          await this.apiCreateNode({
+            ...OsrdNgeSync.castNgeNode(node, netzgrafikDto.labels),
+            path_item_key: key,
+          });
+          // keep track of the ID given by NGE
+          this.nodesIdToKey[node.id] = key;
+        }
+        break;
+      }
+      case 'delete': {
+        if (indexNode) await this.apiDeleteNode(indexNode.node);
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  /**
+   * Check if we have duplicates
+   * Ex: one key is trigram and an other is uic (with the same trigram), we should keep trigram
+   * What we do :
+   *  - Make a list of key,trigram
+   *  - aggregate on trigram to build a list of key
+   *  - filter if the array is of size 1 (ie, no dedup todo)
+   *  - sort the keys by priority
+   *  - add redirection in the nodesByPathKey
+   */
+  private dedupNodes(): void {
+    const trigramAggreg = Object.entries(this.nodesByPathKey)
+      .filter(([_, value]) => typeof value !== 'string' && value.node.trigram)
+      .map(([key, value]) => ({ key, trigram: (value as NodeIndex).node.trigram }))
+      .reduce(
+        (acc, curr) => {
+          acc[`${curr.trigram}`] = [...(acc[`${curr.trigram}`] || []), curr.key];
+          return acc;
+        },
+        {} as Record<string, string[]>
+      );
```

Comment:
- you can add a `!` after `trigram` since you know it is defined
- remove the strings l.323
```suggestion
      .map(([key, value]) => ({ key, trigram: (value as NodeIndex).node.trigram! }))
      .reduce(
        (acc, curr) => {
          acc[curr.trigram] = [...(acc[curr.trigram] || []), curr.key];
          return acc;
        },
        {} as Record<string, string[]>
      );
```